src/Admin/PairedBrowsing.php

```diff
@@ -136,7 +136,7 @@ public function filter_validated_url_status_actions( $actions, WP_Post $post ) {
 	 * Initialize frontend.
 	 */
 	public function init_frontend() {
-		if ( ! amp_is_available() || ! amp_is_dev_mode() ) {
+		if ( ! amp_is_available() || ! amp_is_dev_mode() || ! is_user_logged_in() ) {
```

Comment:
@pierlon The `is_user_logged_in()` condition is being added here and duplicated twice in `MobileRedirection`, which I'm not happy about. I played with extracting that into a helper method of `PairedBrowsing` in https://github.com/ampproject/amp-wp/pull/6526. Curious about your thoughts given how this essentially makes it a `Conditional`-“light” service.

Reply:
How about adding `amp_is_dev_mode() && is_user_logged_in()` to `is_needed()`, and then in `MobileRedirection` check to see if the `admin.paired_browsing` service exists in the service container via `Services::has('admin.paired_browsing')`.

With that implemented, in #6526 you could then remove `PairedBrowsing` from the `MobileRedirection` constructor, and replace instances of `$this->paired_browsing->is_available()` with `Services::has('admin.paired_browsing')`.

cc @schlessera

Reply:
That's a good thought. I wondered about putting those checks inside `is_needed` but I didn't think it would work because `is_user_logged_in()` needs to run at or later than `set_current_user`. It seems this is doable by making the service `Delayed` however, and using `init` or `wp_loaded` as the action. I'm giving it a try.

Reply:
OK, that works pretty well! See 35bed57. The only problem is with testing. Namely, `MobileRedirectionTest` fails in `test_is_using_client_side_redirection` and `test_add_mobile_version_switcher` because the `PairedBrowsing` service isn't registered anymore. Any thoughts on that?

Reply:
I've got something working to test, although it doesn't feel _super_ elegant given the private access: 8097ac9aa. But it seems to work well.

Reply:
>  Namely, MobileRedirectionTest fails in test_is_using_client_side_redirection and test_add_mobile_version_switcher because the PairedBrowsing service isn't registered. Any thoughts on that?

Triggering the `wp_loaded` action should cause the PairedBrowsing service to be registered and thus the tests to pass, but that does not seem to be the case for some reason.

Reply:
I think it's because the registration of services is aborted once there is one service registered:

https://github.com/ampproject/amp-wp/blob/eef6a5f4129c0f5f321d17e66efe0ddd5a409381/src/Infrastructure/ServiceBasedPlugin.php#L156-L160

Reply:
That count should be 0 since a new service container is initialized before each test run:

https://github.com/ampproject/amp-wp/blob/b1151015c343b4755e01ee4d7bdb618aef0c037f/tests/php/src/DependencyInjectedTestCase.php#L45-L46

Reply:
But that registration happens at `setUp` before the test runs, so by the time the test tries doing anything the service container is already populated.

Reply:
Right, and at that point the registration of the `PairedBrowsing` service would have been delayed to occur at the `wp_loaded` action.

It seems that the service container returned by the `Services` helper is not the same instance as `$this->container` in `DependencyInjectedTestCase`. Commenting the line that overrides `self::container` in the `Services` helper will cause the service container to be retrieved from `$this->plugin` in `DependencyInjectedTestCase`, and that seems to work:

```diff
diff --git a/tests/php/src/MobileRedirectionTest.php b/tests/php/src/MobileRedirectionTest.php
--- a/tests/php/src/MobileRedirectionTest.php	(revision 35bed57881c7782bfe660a8f24289f250f595771)
+++ b/tests/php/src/MobileRedirectionTest.php	(date 1628632586105)
@@ -2,6 +2,7 @@
 
 namespace AmpProject\AmpWP\Tests;
 
+use AmpProject\AmpWP\Admin\PairedBrowsing;
 use AmpProject\AmpWP\Admin\ReaderThemes;
 use AmpProject\AmpWP\Infrastructure\Registerable;
 use AmpProject\AmpWP\Infrastructure\Service;
@@ -9,6 +10,7 @@
 use AmpProject\AmpWP\PairedRouting;
 use AmpProject\AmpWP\QueryVar;
 use AmpProject\AmpWP\MobileRedirection;
+use AmpProject\AmpWP\Services;
 use AmpProject\AmpWP\Tests\Helpers\AssertContainsCompatibility;
 use AMP_Options_Manager;
 use AMP_Theme_Support;
@@ -429,10 +431,17 @@
 
 		wp_set_current_user( self::factory()->user->create( [ 'role' => 'administrator' ] ) );
 		add_filter( 'amp_dev_mode_enabled', '__return_true' );
+
+		// Trigger `PairedBrowsing` service to be registered.
+		do_action('wp_loaded');
+
 		$this->assertTrue( $this->instance->is_using_client_side_redirection() );
 		remove_filter( 'amp_dev_mode_enabled', '__return_true' );
 		wp_set_current_user( 0 );
 
+		// Remove `PairedBrowsing` service from the service container. Attempting to
+       // remove it via `$this->container->offsetUnset( 'admin.paired_browsing' )` doesn't work.
+		Services::get_container()->offsetUnset( 'admin.paired_browsing' );
+
 		$this->assertFalse( $this->instance->is_using_client_side_redirection() );
 		global $wp_customize;
 		require_once ABSPATH . 'wp-includes/class-wp-customize-manager.php';
diff --git a/tests/php/src/DependencyInjectedTestCase.php b/tests/php/src/DependencyInjectedTestCase.php
--- a/tests/php/src/DependencyInjectedTestCase.php	(revision 35bed57881c7782bfe660a8f24289f250f595771)
+++ b/tests/php/src/DependencyInjectedTestCase.php	(date 1628632830675)
@@ -51,7 +51,10 @@
 		// The static Services helper has to be modified to use the same objects
 		// as the ones that are injected into the tests.
 		$this->set_private_property( Services::class, 'plugin', $this->plugin );
-		$this->set_private_property( Services::class, 'container', $this->container );
+
+//		Let the Services helper retrieve the container directly from `$this->plugin`.
+//		$this->set_private_property( Services::class, 'container', $this->container );
+
 		$this->set_private_property( Services::class, 'injector', $this->injector );
 	}
 
```

Reply:
Ah nevermind, uncommenting that line would mean `Services::$container` is the original service container that was created when the plugin was being bootstrapped. The issue then would be why is the `PairedBrowsing` service being registered in the original service container and not the one created for the test?

Reply:
Is not the issue that `PairedBrowsing` is not being registered at all because its `is_needed` method returns `false` during `setUp`?

Reply:
Oh right, I forgot to add this line just before calling `do_action('wp_loaded')`:

`AMP_Options_Manager::update_option( Option::THEME_SUPPORT, AMP_Theme_Support::TRANSITIONAL_MODE_SLUG );`

Which would then cause `is_needed` to return true.